[flang] fix build on FreeBSD after 4762c6557d15

brooksdavis opened the pull request:
The HUGE definition collides with the HUGE macro from math.h.  Unlike the fix in 3149c934cb26 (#84478) (largely reverted in f95710c76519), add another #undef HUGE since there is no practical way to make FreeBSD's headers not define HUGE and still define XSI interfaces such as isascii or strnlen.

Fixes #86038

brooksdavis committed:
[flang] fix build on FreeBSD after 4762c6557d15

The HUGE definition collides with the HUGE macro from math.h.  Unlike
the fix in 3149c934cb26 (#84478) (largely reverted in f95710c76519), add
another #undef HUGE since there is no practical way to make FreeBSD's
headers not define HUGE and still define XSI interfaces such as isascii or
strnlen.

Fixes #86038

luporl commented on flang/lib/Decimal/decimal-to-binary.cpp:
```diff
@@ -16,6 +16,10 @@
 #include <ctype.h>
 #include <utility>
 
+// Some environments, viz. glibc 2.17, allow the macro HUGE
+// to leak out of <math.h>.
```
Please mention in this comment that FreeBSD also allows the macro HUGE to leak out of <math.h>.

brooksdavis replied:
I've added a second commit to update all the comments.

luporl reviewed (approved):
If there is no practical way to make FreeBSD's headers not define HUGE and still define XSI interfaces such as isascii or strnlen, then there's not much choice but to undef HUGE.

A possibly better option would be to include `<cmath>` and undef HUGE in `flang/lib/Decimal/big-radix-floating-point.h`, to avoid issues with other sources using its HUGE definition in the future. By doing this in every header file that declares a HUGE symbol, it should be possible to avoid hitting this issue again.

brooksdavis committed:
[flang] *BSD leak HUGE macro

Update comments above `#undef HUGE` to reflect the fact that all major
BSD versions (I checked DragonFly, FreeBSD, NetBSD, and OpenBSD) leak
the HUGE macro from math.h to various degrees.

brooksdavis commented:
I'll leave a better fix to the flang developers. It appears that HUGE is de facto part of the math.h interface and POSIX should just define it if the definitions are reasonably well aligned.
